[17.0][MIG] cetmix_tower_server_queue: forward port updates

Kurama2d opened the pull request:
Forward port the following PR:
[[16.0][MIG] cetmix_tower_server_queue: forward port updates](https://github.com/cetmix/cetmix-tower/pull/340)

Task: 4840

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced asynchronous file upload and download capabilities with job queue management and user notifications.
  * Added visual status indicators (Processing, Success, Error) for file operations in both form and list views.
  * Added a new field to command logs to track command status, with detailed status codes.

* **Bug Fixes**
  * Improved error handling for file upload and download operations, ensuring proper state management and user feedback.

* **Tests**
  * Added comprehensive tests for asynchronous file transfer operations, including error scenarios and state management.
  * Expanded test coverage for command log job failure handling.

* **Documentation**
  * Updated feature descriptions and release notes to reflect new asynchronous file transfer functionality.

* **Chores**
  * Updated module manifests and imports to include new models, views, and test modules.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Kurama2d commented on cetmix_tower/readme/newsfragments/3720.feature:
```diff
@@ -0,0 +1 @@
+cetmix_tower_server_queue: Add async file upload/download via job queue
```
It's a PR with migration to higher version, no changes needed

Kurama2d commented on cetmix_tower_server_queue/tests/test_file.py:
```diff
@@ -0,0 +1,196 @@
+
+            self.assertEqual(len(trap.enqueued_jobs), 2)
+
+            upload_file.write({"server_response": "ok", "is_being_processed": False})
+            upload_file_2.write({"server_response": "ok", "is_being_processed": False})
+
```
It's a PR with migration to higher version, no changes needed

Kurama2d commented on cetmix_tower_server_queue/tests/test_command_log.py:
```diff
@@ -0,0 +1,43 @@
+    # Should match models.queue_job.QUEUE_JOB_ERROR
+    QUEUE_JOB_ERROR = 601
+
+    def test_command_log_state_on_job_fail(self):
+        command = self.env["cx.tower.command"].create(
+            {
+                "name": "Test Command",
+                "action": "ssh_command",
+                "code": "echo 'Hello World'",
+            }
+        )
+        self.assertTrue(command.id, "Command should be created successfully")
+
+        self.server_test_1.run_command(command=command)
+        command_log = self.env["cx.tower.command.log"].search(
+            [("command_id", "=", command.id)], order="id desc", limit=1
+        )
+        self.assertTrue(command_log, "Command log should be created")
+
+        job = command_log.queue_job_id
+        self.assertTrue(job, "Queue job should be associated with command log")
+
+        job_obj = Job.load(self.env, job.uuid)
+        job_obj.set_failed()
+        job_obj.store()
+        self.assertEqual(job.state, "failed", "Job should be in failed state")
+        self.assertEqual(
+            command_log.command_status,
+            self.QUEUE_JOB_ERROR,
+            "Command log should be in failed state",
+        )
```
It's a PR with migration to higher version, no changes needed

Kurama2d commented on cetmix_tower_server_queue/tests/test_command_log.py:
```diff
@@ -0,0 +1,43 @@
+    def setUp(self):
+        super().setUp()
+
+    # Should match models.queue_job.QUEUE_JOB_ERROR
+    QUEUE_JOB_ERROR = 601
```
It's a PR with migration to higher version, no changes needed

Kurama2d commented on cetmix_tower_server_queue/tests/__init__.py:
```diff
@@ -1 +1,3 @@
 from . import test_command
+from . import test_command_log
+from . import test_file
```
It's a PR with migration to higher version, no changes needed

Kurama2d commented on cetmix_tower_server_queue/readme/newsfragments/4718.feature:
```diff
@@ -0,0 +1 @@
+Terminate command with error if job has failed
```
It's a PR with migration to higher version, no changes needed

Kurama2d commented on cetmix_tower_server_queue/models/queue_job.py:
```diff
@@ -0,0 +1,23 @@
+
+    QUEUE_JOB_ERROR = 601
+
+    def write(self, vals):
+        """
+        Override write method to update command status
+        and write error information in the log record
+        """
+        if vals.get("state") == "failed":
+            log_record = self.kwargs.get("log_record")
+            if log_record:
+                log_record.finish(
+                    status=self.QUEUE_JOB_ERROR,
+                    error=vals.get("exc_info"),
+                )
+        return super().write(vals)
```
It's a PR with migration to higher version, no changes needed

Kurama2d commented on cetmix_tower_server_queue/models/queue_job.py:
```diff
@@ -0,0 +1,23 @@
+class QueueJob(models.Model):
+    _inherit = "queue.job"
+
+    QUEUE_JOB_ERROR = 601
```
It's a PR with migration to higher version, no changes needed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,162 @@
+        finally:
+            self.write({"is_being_processed": False})
+
+    def _do_download(self, raise_error=True):
+        """
+        Downloads the file within a job context and notifies the user on success.
+        Logs the error if an exception occurs;
+        failure state is managed by the parent method.
+        """
+        try:
+            with self.env.cr.savepoint():
+                result = super().download(raise_error=raise_error)
+                single_msg = _("File downloaded!")
+                plural_msg = _("Files downloaded!")
+                self.env["bus.bus"].sudo().sendone(
+                    (self._cr.dbname, "res.partner", self.env.user.partner_id.id),
+                    {
+                        "message": single_msg if len(self) == 1 else plural_msg,
+                        "title": _("Success"),
+                        "type": "success",
+                    },
+                )
+                return result
+        except Exception as e:
+            if not raise_error:
+                _logger.error("File %s download failed: %s", str(self), str(e))
+            else:
+                raise
+        finally:
+            self.write({"is_being_processed": False})
```
It's a PR with migration to higher version, no changes needed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,162 @@
+                result = super().upload(raise_error=raise_error)
+                single_msg = _("File uploaded!")
+                plural_msg = _("Files uploaded!")
+                self.env["bus.bus"].sudo().sendone(
+                    (self._cr.dbname, "res.partner", self.env.user.partner_id.id),
+                    {
+                        "message": single_msg if len(self) == 1 else plural_msg,
+                        "title": _("Success"),
+                        "type": "success",
+                    },
+                )
```
It's a PR with migration to higher version, no changes needed

Kurama2d commented on cetmix_tower_server_queue/models/cx_tower_file.py:
```diff
@@ -0,0 +1,162 @@
+                raise
+        finally:
+            self.write({"is_being_processed": False})
+
```
It's a PR with migration to higher version, no changes needed

GabbasovDinar commented on cetmix_tower_server_queue/views/cx_tower_file_view.xml:
```diff
@@ -2,33 +2,33 @@
-        <field name="name">cx.tower.file.view.form</field>
-        <field name="model">cx.tower.file</field>
+    <field name="name">cx.tower.file.view.form</field>
+    <field name="model">cx.tower.file</field>
```
incorrect file formating, please fix.

```xml
<record id="......">
    <field name="">...</field>
```

Kurama2d replied:
Thanks, fixed it.

Kurama2d commented on cetmix_tower_server_queue/views/cx_tower_file_view.xml:
```diff
@@ -0,0 +1,56 @@
+                </attribute>
+                <attribute name="decoration-danger">
+                    is_being_processed != True and server_response not in ('ok', False)
+                </attribute>
```
Fixed it.

Aldeigja reviewed (approved):
functional LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

Aldeigja reviewed (approved):
Functional LGTM

ivs-cetmix committed:
[IMP] cetmix_tower_server Prep for async file sync

- Add error codes for file sync.
- Do not copy response value when file is duplicated.

Task: 3720

ivs-cetmix committed:
[IMP] cetmix_tower_server_queue Async file ops

Implement queue_job for file operations.
Terminate command with error job has failed.

Task: 3720
Task: 4718

ivs-cetmix commented:
/ocabot merge minor

CetmixGitDrone commented:
On my way to merge this fine PR!
Prepared branch [17.0-ocabot-merge-pr-354-by-ivs-cetmix-bump-minor](https://github.com/cetmix/cetmix-tower/commits/17.0-ocabot-merge-pr-354-by-ivs-cetmix-bump-minor), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 2beff39c5ee1cc71bd0effbb9a12432d30266094. Thanks a lot for contributing to cetmix. ❤️
